GDJS/Runtime/variable.ts

```diff
@@ -211,73 +285,205 @@ namespace gdjs {
      * @param newValue The new string to be set
      */
     setString(newValue: string): void {
-      this._str = newValue;
-      this._numberDirty = true;
-      this._stringDirty = false;
+      this._type = 'string';
+      this._str = '' + newValue;
+    }
+
+    /**
+     * Get the value of the variable, considered as a boolean
+     * @return {boolean}
+     */
+    getAsBoolean(): boolean {
+      if (this._type !== 'boolean') {
+        if (this._type === 'number') return this._value !== 0;
+        else if (this._type === 'string')
+          return this._str !== '0' && this._str !== '' && this._str !== 'false';
+        else return this.getChildrenCount() !== 0;
+      }
+
+      return this._bool;
+    }
+
+    /**
+     * Change the value of the variable, considered as a boolean
+     * @param {boolean} newValue
+     */
+    setBoolean(newValue: boolean) {
+      this._type = 'boolean';
+      this._bool = !!newValue;
+    }
+
+    /**
+     * Sets the primitive value using the setter of the current type.
+     * @param {any} newValue
+     */
+    setValue(newValue: string | float | boolean) {
+      if (this._type === 'string') this.setString(newValue as string);
+      else if (this._type === 'number') this.setNumber(newValue as float);
+      else if (this._type === 'boolean') this.setBoolean(newValue as boolean);
+    }
+
+    /**
+     * Gets the primitive value using the getter of the current type.
+     * @param {any} newValue
+     */
+    getValue(): string | float | boolean {
+      return this._type === 'number'
+        ? this.getAsNumber()
+        : this._type === 'boolean'
+        ? this.getAsBoolean()
+        : this.getAsString();
     }
 
     /**
      * Return true if the variable is a structure.
      * @return true if the variable is a structure.
+     * @deprecated Use {@link gdjs.Variable.getType} instead.
      */
     isStructure(): boolean {
-      return this._isStructure;
+      return this._type === 'structure';
     }
 
     /**
      * Return true if the variable is a number.
      * @return true if the variable is a number.
+     * @deprecated Use {@link gdjs.Variable.getType} instead.
      */
     isNumber(): boolean {
-      return !this._isStructure && !this._numberDirty;
+      return this._type === 'number';
+    }
+
+    /**
+     * Returns the type of the variable.
+     * @return {VariableType}
+     */
+    getType(): VariableType {
+      return this._type;
     }
 
     /**
      * Return the object containing all the children of the variable
      * @return All the children of the variable
      */
-    getAllChildren(): { [key: string]: gdjs.Variable } {
-      return this._children;
+    getAllChildren(): Children {
+      return this._type === 'structure'
+        ? this._children
+        : this._type === 'array'
+        ? ((Object.assign({}, this._childrenList) as unknown) as Children)
```

Comment:
What is this thing doing? Should we rather return an empty object because if you call `getAllChildren` on an array, you're misusing the API anyway?

Reply:
The getters implicitly convert the value to the good type, so I think it would be better to stay consistent and do it everywhere. This converts the array to an object by taking the properties (indices) and assigning them on the empty object.